### PR TITLE
S2259: Support member access

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
~Sonar and Roslyn versions don't run completely independently at the moment. Should be solved with #5911.~

Update #5911 is merged and solved the problem.

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 1]
```diff
@@ -119,7 +119,8 @@ void NonCompliant2()
       switch (o?.GetHashCode())
       {
         case null:
-          o.ToString(); // FIXME Non-compliant
+          o.ToString(); // Noncompliant
+                        // Noncompliant@-1 FIXME: double reporting
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:123]: Seems to be related to switch branching. Using `PostProcessSimple` does not change behavior here.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/NullPointerDereferenceBase.cs [hunk 2]
```diff
@@ -18,10 +18,33 @@
...
 {
     public abstract class NullPointerDereferenceBase : SymbolicRuleCheck
     {
         internal const string DiagnosticId = "S2259";
+
+        protected override ProgramState PreProcessSimple(SymbolicContext context)
+        {
+            switch (context.Operation.Instance.Kind)
+            {
+                case OperationKindEx.Invocation:
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/NullPointerDereferenceBase.cs:36]: A second operationkind is underway: 903d9a110392345810c11b5940f8434450c8987a

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs [hunk 3]
```diff
@@ -670,7 +669,7 @@ private void Invoke()
         {
             this.bar = null;
             if (this.bar != null)
-                this.bar.GetHashCode();
+                this.bar.GetHashCode(); // Noncompliant FIXME (was compliant before)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:672]: Please rebase, this should go away and could break master

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
That's a very nice babystep

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/NullPointerDereferenceBase.cs [hunk 4]
```diff
@@ -18,10 +18,33 @@
...
+                    if (IInvocationOperationWrapper.FromOperation(context.Operation.Instance) is { } invocation
+                        && invocation.Instance is { } instance
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/NullPointerDereferenceBase.cs:38]: Returns null for static methods.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 5]
```diff
@@ -52,11 +53,11 @@ public void TryCatch4(object o)
             o = null;
             try
             {
-                var a = o?.ToString();
+                var a = o?.ToString();  // Compliant
             }
             catch (Exception e) when (e.Message != null)
             {
-                var b = o.ToString(); // FIXME Non-compliant, o could be null here
+                var b = o.ToString(); // Compliant FIXME (was Non-compliant before)
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:60]: My suspicion is, that the engine is smart enough to see that the catch block is unreachable (There is no code path that can throw). But maybe I'm wrong.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 6]
```diff
@@ -13,7 +13,8 @@ void ConditionalThisFieldAccess()
         {
             object o = null;
             this._foo1 = o;
-            this?._foo1.ToString(); // FIXME Non-compliant
+            this?._foo1.ToString();
+            //   ^^^^^^              {{'._foo1' is null on at least one execution path.}}
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:17]: Notice the `.` in the message. It is produced by the `instance.Syntax.ToString()`

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicContext.cs [hunk 7]
```diff
@@ -43,5 +43,14 @@ public SymbolicContext(IOperationWrapperSonar operation, ProgramState state)
...
             State == newState ? this : new(Operation, newState);
+
+        public bool HasConstraint(SymbolicConstraint constraint) =>
+            HasConstraint(constraint, Operation);
+
+        public bool HasConstraint(SymbolicConstraint constraint, IOperationWrapperSonar operation) =>
+            HasConstraint(constraint, operation.Instance);
+
+        public bool HasConstraint(SymbolicConstraint constraint, IOperation operation) =>
+            State[operation]?.HasConstraint(constraint) is true;
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicContext.cs:54]: Maybe it is better to move these "helpers" into some extension methods class.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicContext.cs:54]: No, that would just scatter the logic over multiple places. We can modify this class, so we don't need extensions.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Final polishing

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs (hunk 5 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:60]: An `Console.WriteLine("");` did the trick.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs [hunk 8]
```diff
@@ -480,28 +479,28 @@ void OtherInstanceFieldAccess2()
         {
             object o = null;
             _foo1 = o;
-            (new NullPointerDereferenceWithFields())._foo1.ToString(); // Compliant
+            (new NullPointerDereferenceWithFields())._foo1.ToString(); // Noncompliant
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:482]: I think it was an FN before. `_foo1` is null here. The invocation instance is a FieldReferenceOperation and the referenced field `_foo1` has the `ObjectConstraint.Null`. I added marker for the syntax. (It is the whole SimpleMemberAccessExpression and not just the Name part).

[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:482]: What is fishy, but not covered by the test is this case:
```cs
            object o = null;
            _foo1 = new object(); // see the difference here.
            (new NullPointerDereferenceWithFields())._foo1.ToString(); // Should raise also, but does not.
```

Should I add this as latest case too?

[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs]: I added `OtherInstanceFieldAccess3`.

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs]: The key here is, that we are not field-sensitive. We should not be able to detect any of those cases

[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs]: > The key here is, that we are not field-sensitive

You mean instance sensitive, don't you (_foo1 of the two instances are confounded)? There is nothing I can do on my side right now, isn't there?

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs]: Yes, instance sensitive.
The problem is not in this PR, there's something unexpected going on elsewhere. We'll take a look later when hunting FIXMEs

**Pull request merged**